### PR TITLE
kbuildbarn: cut down api, add files and make more functional

### PR DESCRIPTION
**Pull request by aaahrens**
Does the following things:
* makes kkbuildbarn use more options and less eLOC. makes it so that url generation is sane.
* Tests and Develops the api for generating files

### CODE REVIEW

File: lib/kbuildbarn/urls.go [hunk 1]
```diff
@@ -145,9 +68,12 @@ func readAndClose(rc io.ReadCloser) ([]byte, error) {
...
 // RetryUntilSuccess just blasts through all possible urls until it hits one that works. This is intended for
 // applications that are blind to the type of artifact
-func RetryUntilSuccess(params BuildBarnParams) ([]byte, error) {
+func RetryUntilSuccess(baseName, hash, size string) ([]byte, error) {
 	urls := []string{
-		params.FileUrl(), params.ActionUrl(), params.CommandUrl(), params.DirectoryUrl(),
+		Url(baseName, hash, size, WithActionUrlTemplate()),
+		Url(baseName, hash, size, WithDirectoryUrlTemplate()),
+		Url(baseName, hash, size, WithCommandUrlTemplate()),
+		Url(baseName, hash, size, WithByteStreamTemplate()),
```
Comment (gpaussaenf) [inline lib/kbuildbarn/urls.go:76]: What about the file:// template? Should it be in this list too, or am I missing something?

**Review by gpaussaenf — APPROVED** [review]
I'm approving these changes, but I had some trouble deciphering how it all works. I'm sure I'll figure it out in time as I start working with these APIs.

File: lib/kbuildbarn/urls.go (hunk 1 above)
Comment (aaahrens) [reply lib/kbuildbarn/urls.go:76]: The File() endpoint ignores the scheme. If you wanted a file://<>/<>, you can use 
`kbuildbarn.WithFooTemplate(), kbuildbarn.WithScheme("file")`

File: lib/kbuildbarn/urls.go [hunk 2]
```diff
@@ -7,83 +7,36 @@ import (
...
 
 // the following default values are arbitrary, based on what current works with buildbarn
-var (
+const (
 	DefaultFileTemplate       = "/blobs/file/%s-%s/%s"
 	DefaultActionTemplate     = "/blobs/action/%s-%s"
 	DefaultCommandTemplate    = "/blobs/command/%s-%s"
 	DefaultDirectoryTemplate  = "/blobs/directory/%s-%s"
 	DefaultByteStreamTemplate = "/blobs/%s/%s"
 )
```
[the remark this answers is not in the document]
Comment (aaahrens) [reply lib/kbuildbarn/urls.go:38]: ah probably

**Pull request merged**